src/components/filters/filter-panels/range-filter-panel.js

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import React, {useCallback} from 'react';
+import {StyledFilterContent} from 'components/common/styled-components';
+import RangeFilterFactory from 'components/filters/range-filter';
+import FilterPanelHeaderFactory from 'components/side-panel/filter-panel/filter-panel-header';
+import FieldSelector from 'components/common/field-selector';
+import SourceDataSelectorFactory from 'components/side-panel/common/source-data-selector';
+
+RangeFilterPanelFactory.deps = [
+  FilterPanelHeaderFactory,
+  RangeFilterFactory,
+  SourceDataSelectorFactory
+];
+
+function RangeFilterPanelFactory(
+  FilterPanelHeader,
+  RangeFilter,
+  SourceDataSelector
+) {
+  const RangeFilterPanel = React.memo(({
+    idx,
+    datasets,
+    allAvailableFields,
+    filter,
+    isAnyFilterAnimating,
+    enlargeFilter,
+    removeFilter,
+    setFilter,
+    toggleAnimation
+  }) => {
+    const onFieldSelector = useCallback(field =>
+        setFilter(idx, 'name', field.name),
+      [idx, setFilter]
+    );
+
+    const onSourceDataSelector = useCallback(value =>
+        setFilter(idx, 'dataId', value),
+      [idx, setFilter]
+    );
+
+    const onSetFilter = useCallback(value =>
+        setFilter(idx, 'value', value),
+      [idx, setFilter]);
+
+    return (
+      <div>
+        <FilterPanelHeader
```

Comment:
```js
const onFieldSelector = useCallback(field =>
        setFilter(idx, 'name', field.name),
      [idx, setFilter]
    );

    const onSourceDataSelector = useCallback(value =>
        setFilter(idx, 'dataId', value),
      [idx, setFilter]
    );

    const onSetFilter = useCallback(value =>
        setFilter(idx, 'value', value),
      [idx, setFilter]);

<PanelHeader>
   <FieldSelector/>
</PanelHeader>
<StyledFilterContent>
    <SourceDataSelecto/>
</StyledFilterContent>
```

Has been repeated in every `xx-filter-panel` component, except polygon filter. Repeating them 5 times, doesn't make sense. Why not make a component `FilterPanelWithFieldSelect` and use it to wrap other filters?

Reply:
I think I decided to go down this way because both polygon and time range filters have panel actions and by simply using children couldn't be recreated but I think we can create the common component and
- panel actions: will be passed as an array of object: `
``{onClick: enlargeFilter, tooltip:: 'Time Playback, iconComponent: Clock, onSelect: 
onFieldSelector}```
- the actual field component, e.g.TimeRangeFilter, will be passed as one of the children